src/envoy/tcp/metadata_exchange/metadata_exchange.cc

```diff
@@ -0,0 +1,260 @@
+      conn_state_ = Done;
+      return Network::FilterStatus::Continue;
+  }
+
+  return Network::FilterStatus::Continue;
+}
+
+Network::FilterStatus MetadataExchangeFilter::onNewConnection() {
+  return Network::FilterStatus::Continue;
+}
+
+Network::FilterStatus MetadataExchangeFilter::onWrite(Buffer::Instance&, bool) {
+  switch (conn_state_) {
+    case Invalid:
+    case Done:
+      // No work needed if connection state is Done or Invalid.
+      return Network::FilterStatus::Continue;
+    case ConnProtocolNotRead: {
+      if (read_callbacks_->connection().nextProtocol() != config_->protocol_) {
+        conn_state_ = Invalid;
+        config_->stats().alpn_protocol_not_found_.inc();
+        return Network::FilterStatus::Continue;
+      } else {
+        conn_state_ = WriteMetadata;
+        config_->stats().alpn_protocol_found_.inc();
+      }
+    }
+    case WriteMetadata: {
+      // TODO(gargnupur): Try to move this just after alpn protocol is
+      // determined and first onWrite is called in Upstream filter.
+      writeNodeMetadata();
+    }
+    case ReadingInitialHeader:
+    case ReadingProxyHeader:
+    case NeedMoreDataInitialHeader:
+    case NeedMoreDataProxyHeader:
+      // These are to be handled in Reading Pipeline.
+      return Network::FilterStatus::Continue;
+  }
+
+  return Network::FilterStatus::Continue;
+}
+
+void MetadataExchangeFilter::writeNodeMetadata() {
+  if (conn_state_ != WriteMetadata) {
+    return;
+  }
+
+  std::unique_ptr<const google::protobuf::Struct> metadata =
+      getMetadata(config_->node_metadata_id_);
+  if (metadata != nullptr) {
+    Envoy::ProtobufWkt::Any metadata_any_value;
+    *metadata_any_value.mutable_type_url() = StructTypeUrl;
+    *metadata_any_value.mutable_value() = metadata->SerializeAsString();
+    std::unique_ptr<::Envoy::Buffer::OwnedImpl> buf =
+        constructProxyHeaderData(metadata_any_value);
+    write_callbacks_->injectWriteDataToFilterChain(*buf, false);
+
+    if (config_->filter_direction_ == FilterDirection::Downstream) {
+      setMetadata(DownstreamDynamicDataKey, *metadata);
+    } else {
+      setMetadata(UpstreamDynamicDataKey, *metadata);
+    }
+    config_->stats().metadata_added_.inc();
+  }
+
+  conn_state_ = ReadingInitialHeader;
+}
+
+void MetadataExchangeFilter::tryReadInitialProxyHeader(Buffer::Instance& data) {
+  if (conn_state_ != ReadingInitialHeader &&
+      conn_state_ != NeedMoreDataInitialHeader) {
+    return;
+  }
+  const uint32_t initial_header_length = sizeof(MetadataExchangeInitialHeader);
+  if (data.length() < initial_header_length) {
+    config_->stats().initial_header_not_found_.inc();
+    // Not enough data to read. Wait for it to come.
+    conn_state_ = NeedMoreDataInitialHeader;
+    return;
+  }
+  std::string initial_header_buf = std::string(
+      static_cast<const char*>(data.linearize(initial_header_length)),
+      initial_header_length);
+  const MetadataExchangeInitialHeader* initial_header =
```

Comment:
But shouldn't marking MetadataExchangeInitialHeader struct as packed attribute help with misalignment errors?

Reply:
No packed struct doesn't mitigate misalignment errors, because you're doing `reinterpret_cast` from a raw pointer which is not guaranteed to be aligned.

Since you're already doing a copy here, why not just do:
`
MetadataExchangeInitialHeader initial_header;
data.copyOut(0, initial_header_length, &initial_header);
`
then where you copied to is aligned.